gax-httpjson/src/main/java/com/google/api/gax/httpjson/HttpRequestRunnable.java

```diff
@@ -109,30 +110,32 @@ public void run() {
       }
 
       // Populate HTTP path and query parameters.
-      Map<String, String> pathParams =
-          requestBuilder.getPathParams(request, methodDescriptor.getResourceNameField());
-      PathTemplate pathPattern = PathTemplate.create(methodDescriptor.endpointPathTemplate());
-      String relativePath = pathPattern.instantiate(pathParams);
-      GenericUrl url = new GenericUrl(endpoint + relativePath);
-      Map<String, List<String>> queryParams =
-          requestBuilder.getQueryParams(request, methodDescriptor.getQueryParams());
-      for (String queryParam : methodDescriptor.getQueryParams()) {
-        if (queryParams.containsKey(queryParam) && queryParams.get(queryParam) != null) {
-          url.set(queryParam, queryParams.get(queryParam));
+      GenericUrl url = new GenericUrl(endpoint + requestFormatter.getEndpointRelativePath(request));
+      Map<String, List<String>> queryParams = requestFormatter.getQueryParams(request);
+      for (String queryParam : queryParams.keySet()) {
+        if (queryParams.get(queryParam) != null) {
+          for (String val : queryParams.get(queryParam)) {
+            url.set(queryParam, val);
+          }
         }
       }
 
       HttpRequest httpRequest =
-          requestFactory.buildRequest(methodDescriptor.getHttpMethod(), url, jsonHttpContent);
+          requestFactory.buildRequest(requestFormatter.getHttpMethod(), url, jsonHttpContent);
       for (HttpJsonHeaderEnhancer enhancer : headerEnhancers) {
         enhancer.enhance(httpRequest.getHeaders());
       }
       httpRequest.setParser(new JsonObjectParser(jsonFactory));
 
       HttpResponse httpResponse = httpRequest.execute();
 
-      ResponseT response =
-          methodDescriptor.parseResponse(new InputStreamReader(httpResponse.getContent()));
+      if (!httpResponse.isSuccessStatusCode()) {
+        throw new ApiException(
+            null,
+            HttpJsonStatusCode.of(httpResponse.getStatusCode(), httpResponse.getStatusMessage()),
+            false);
```

Comment:
Shouldn't we use `ApiExceptionFactory` so we can get the right subclass of `ApiException`?

Reply:
Yep!